### PR TITLE
Migrate Parser to PHP 8

### PR DESCRIPTION
**Pull request by greg0ire**
In some cases, I had to widen the documented type, running the test suite showed me it was wrong.

### CODE REVIEW

File: lib/Doctrine/ORM/Query/Parser.php [hunk 1]
```diff
@@ -2541,18 +2354,8 @@ public function ConditionalPrimary()
...
-     *         AST\EmptyCollectionComparisonExpression|
-     *         AST\ExistsExpression|
-     *         AST\InExpression|
-     *         AST\InstanceOfExpression|
-     *         AST\LikeExpression|
-     *         AST\NullComparisonExpression
-     */
-    public function SimpleConditionalExpression()
+     */
+    public function SimpleConditionalExpression(): AST\ExistsExpression|AST\BetweenExpression|AST\LikeExpression|AST\InListExpression|AST\InSubselectExpression|AST\InstanceOfExpression|AST\CollectionMemberExpression|AST\NullComparisonExpression|AST\EmptyCollectionComparisonExpression|AST\ComparisonExpression
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Query/Parser.php:2358]: Should I simplify this one to `AST\Node`?

File: lib/Doctrine/ORM/Query/Parser.php [hunk 2]
```diff
@@ -370,19 +334,11 @@ public function parse()
...
-                    break;
-
-                case $AST instanceof AST\SelectStatement:
-                default:
-                    $treeWalkerChain->walkSelectStatement($AST);
-            }
+            match (true) {
+                $AST instanceof AST\UpdateStatement => $treeWalkerChain->walkUpdateStatement($AST),
+                $AST instanceof AST\DeleteStatement => $treeWalkerChain->walkDeleteStatement($AST),
+                default => $treeWalkerChain->walkSelectStatement($AST),
```
Comment (derrabus) [inline lib/Doctrine/ORM/Query/Parser.php]: Let's be explicit about the remaining case that we expect here. Static analysis should recognize that the `match` statement covers all cases without a `default` and will notify us if that will ever change in the future. 💪🏻 

```suggestion
                $AST instanceof AST\SelectStatement => $treeWalkerChain->walkSelectStatement($AST),
```

**Review by derrabus — CHANGES_REQUESTED** [review]
Shall we `final`ize the `Parser` class? With all those native union types, it'll otherwise hard to change that method in the future.

File: lib/Doctrine/ORM/Query/Parser.php (hunk 1 above)
Comment (derrabus) [reply lib/Doctrine/ORM/Query/Parser.php:2358]: Depends on how we answer my question regarding `final`. 🙂

**Comment by greg0ire** [comment]
Let's try `final`. If it turns out this was a wrong move, we can always remove it afterwards, after adjusting the return types.

File: lib/Doctrine/ORM/Query/Parser.php (hunk 2 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/Query/Parser.php]: Looks nicer indeed 👍

**Review by derrabus — APPROVED** [review]

**Pull request merged**